src/util/log/fd_log.c

```diff
@@ -628,6 +629,68 @@ fd_log_private_2( int          level,
   abort();
 }
 
+void 
+fd_log_hexdump ( const char * descr,
+                 const void * blob,
+                 const size_t len ) {
+    
+  size_t count, blob_len;
+  int is_truncated = 0, len_written;
+  char line_buf[ FD_LOG_HEXDUMP_BYTES_PER_LINE+1 ];
+  const char * blob_ptr = (const char *)blob;
+  
```

Comment:
There are some notions being kicked around for collecting various discussions and comments throughout the headers into one place as a style guide / common practices / etc but in the meantime, some minor pedantic nit picks (not everything applies to to this code but is closely related and thus might be useful):

- The linguistic machine model guarantees by the development environment guarantees that the target platform will have `ulong` is interchangeable with `size_t` and `size_t` has been a 2nd class citizen in the strict C/C++ specification historically (just a mess in the standards to how it evolved).  So `ulong` is generally preferred over `size_t`.

- Some of this might be aspirational but we try just have one variable in a declaration statement, try to use the suffix `_len` on a variable when it refers the length of c-style string len in a `strlen`-sense, try to use the suffix `_sz` to refer to the actual number of contiguous bytes used in the memory (e.g. `ulong cstr_len = strlen( cstr ); ulong cstr_sz = cstr_len + 1UL;`), try to use the suffix `_cnt` when a variable is a natural number count of something and try to use the suffix `_idx` when a variable is natural number index of something in the half open interval of the corresponding `_cnt` (e.g. `ulong foo_cnt; ulong foo_idx; /* in [0,foo_cnt) */`).  While in a case like this might seem superfluous, it is a bit of a habit to reduce risk of making off-by-one / overflow / underflow / ... errors in more complex cases, increase the chance reviews will catch such errors, make room for commenting domains of individual variables and what not.

- And while not an issue here, note that we also try to use `ulong` for the in-register representation of such.  This is to avoid unpleasant surprises in the language itself around UB/IB behaviors with signed types and narrow types (this area of the C/C++ language specs best described as a superfund site), avoid unpleasant surprises when code gets used in the real world on larger data sets that have been tested in the lab (lost too many weekends debugging some code because there was an `int` iterator hidden somewhere deep in the bowels of 3rd party dependency when a user fired up the code a large data for the first time), to reflect how 64-bit CPUs work under the hood and thus improve quality of compiler code gen, etc.

- The precedence rules in C/C++ for modifiers in variable declarations tend to "just work" naturally if read right-to-left unfortunately (another superfund site in the language).  So another habit we try to do that doesn't make much a difference here but can matter in more nuanced situations with restricted pointers, volatile pointers, etc is to write variable declarations in a way that reads naturally right to left.  E.g. "unmodifiable pointer to a unpredictable pointer to an unpredictable but unmodifiable character" would be a `char const volatile * volatile * const` ... tangentially related, we aspire to avoid writing such types when possible because they aren't exactly the most readable).

- Also not an issue here but the C/C++ language is also a disaster area around the sign of `char` (language leaves it IB and different architectures use different behaviors and the standard library and implementations are a bit of free for all to try to get consistent cross platform behavior around something as elementary as `strcmp` ... look at what the standard says or doesn't and glibc does here for an example).  We try to use `char` for contents of a c-style string and avoid arithmetic on it, `uchar` unsigned 8-bit integral data, `schar` for signed 8-bit integral data and `void *` for pointers to raw binary data.

- We also try to put variable declarations as close to use as possible and then limit the scope of variable declarations to be as narrow as possible.

Combining all this (and the above comment), a possible function call signature might look like:

```
void
fd_log_hexdump( int          level,
                char const * descr,
                void const * blob,
                ulong        sz ) {
  ... move the closer to use and initialize on declaration if possible ...
  ulong blob_sz;
  ...
}
```

Reply:
Thanks very much for these in-depth guidelines - much appreciated. Solid rules and guidelines. I'll rework according to these.